### PR TITLE
Patch VecInit.fill(0) invocation to successfully compile and yield a zero-width Vec

### PR DESCRIPTION
**Pull request by jared-barocsi**
Fixes common workarounds for instantiating a 0-width Vec, like:
```
if(n < 1) Wire(Vec(0, Bool())) else VecInit.fill(n)(false.B)
```
by making a direct `VecInit.fill(0)` call possible.

Some connection semantics issues were discovered however which allow users to connect zero `Vecs` of one type to another zero `Vec` of a different type, because the existing logic only operates on a per-child-element basis and thus doesn't execute at all for `Aggregates` containing no child elements. A temporary fix is applied for this with the additional note that the connection logic needs to be looked at for future unification.
### Contributor Checklist

- [ ] Did you add Scaladoc to every public function/method?
- [x] Did you add at least one test demonstrating the PR?
- [x] Did you delete any extraneous printlns/debugging code?
- [x] Did you specify the type of improvement?
- [x] Did you add appropriate documentation in `docs/src`?
- [x] Did you state the API impact?
- [x] Did you specify the code generation impact?
- [x] Did you request a desired merge strategy?
- [x] Did you add text to be included in the Release Notes for this change?

#### Type of Improvement

- feature
- bug fix

#### API Impact

Modifies `VecInit.fill`

#### Backend Code Generation Impact

No impact

#### Desired Merge Strategy

Squash and merge

#### Release Notes

Fix `VecInit.fill(0)` calls so that they compile and yield 0-width `Vecs`

### Reviewer Checklist (only modified by reviewer)
- [ ] Did you add the appropriate labels?
- [ ] Did you mark the proper milestone (Bug fix: `3.5.x` or `3.6.x` depending on impact, API modification or big change: `5.0.0`)?
- [ ] Did you review?
- [ ] Did you check whether all relevant Contributor checkboxes have been checked?
- [ ] Did you do one of the following when ready to merge:
  - [ ] Squash: You/ the contributor `Enable auto-merge (squash)`, clean up the commit message, and label with `Please Merge`.
  - [ ] Merge: Ensure that contributor has cleaned up their commit history, then merge with `Create a merge commit`.

### CODE REVIEW

**Comment by mwachs5** [comment]
@jackkoenig i think we'd apply the `feature` label here if we can pick only one? The "it's now a bug to do something you should not have been doing" seems OK right?

**Comment by jackkoenig** [comment]
@mwachs5 I agree with `Feature`, `BugFix` is also reasonable but I think `Feature` makes the most sense.

File: src/test/scala/chiselTests/Vec.scala [hunk 1]
```diff
@@ -276,6 +276,56 @@ class VecSpec extends ChiselPropSpec with Utils {
...
+      val chirrtl = emitCHIRRTL(new Module {
+        val io = IO(new Bundle {
+          val out = Vec(0, UInt(8.W))
+        })
+
+        val zeroWidthVec = VecInit.fill(0)(gen)
+
+        connect(io.out, zeroWidthVec)
+      })
+      println(chirrtl)
```
Comment (jackkoenig) [inline src/test/scala/chiselTests/Vec.scala:306]:
```suggestion
```

Comment (jackkoenig) [inline src/test/scala/chiselTests/Vec.scala:306]: Can these tests be combined, basically the bottom seems to be testing the same thing and you could probably do the same chirrtl check on every iteration of the bottom test?

**Review by jackkoenig — APPROVED** [review]

**Pull request merged**